### ISSUE TITLE
It's bug in "Fl_GDI_Graphics_Driver_image.cxx"?

### ISSUE THREAD

**Issue by xuyun018** [fltk/fltk#1033]
In line 215, "to += w;" is no use.
In line 267, When h > blocking, There will be pointer errors, The variable "buffer" will be set to "NULL" and then reused.

**Comment by MatthiasWM** [comment]
https://github.com/fltk/fltk/blob/68f437b63e9e54769b1854097f2c95dd1cce12b7/src/drivers/GDI/Fl_GDI_Graphics_Driver_image.cxx#L215

and

https://github.com/fltk/fltk/blob/68f437b63e9e54769b1854097f2c95dd1cce12b7/src/drivers/GDI/Fl_GDI_Graphics_Driver_image.cxx#L267

for reference

**Comment by MatthiasWM** [comment]
I don't have a Windows machine availabel ATM, so I am somewhat giessing here. 

Line 215 ensures that the pointe to the destination buffer is incremented to the next line. If in indexd mode, colors are stored in a byte sized lookup table. The grayscale driver does the same thing in line 222, and the RGB `to` is incremnetd in line 235, multiplied by the loop across the image width, doing practically the same as line 215.

**Comment by MatthiasWM** [comment]
The buffer that is deleetd in 267 is declared static in line 172: `static U32* buffer;`. I assume that the author deleted the buffer for priniting devices, becuse the RGB pixel data is often a lot bigger than screen data, and he does not want to keep a huge buffer around forever.

It gets reallocated if needed in line 184. This looks ok to me.

I think you may be right that there is an issue if `h > blocking`, but I have not found the location yet. Could you point that out?

**Comment by Albrecht-S** [comment]
@xuyun018 Thanks for the report. I investigated the issue (as @MatthiasWM did) and my conclusion is similar to Matt's result(s):
1. Line 215 `to += w;` is indeed unnecessary because `to` is no longer used after this statement inside the enclosing loop. It is defined (again) in the next iteration at https://github.com/fltk/fltk/blob/68f437b63e9e54769b1854097f2c95dd1cce12b7/src/drivers/GDI/Fl_GDI_Graphics_Driver_image.cxx#L208
2. I don't think that the second issue is a problem. Adding to @MatthiasWM's analysis: `h > blocking` is not possible at this point because this code block deals with printing where blocking is disabled. I agree with Matthias that deleting the buffer is done in this case because the buffer may be very large because printing is done with one large buffer as opposed to the normal "blocking" mode.

Conclusion:
1. the superfluous statement `to += w;` can be removed
2. deleting the buffer is intended and does no harm.

@MatthiasWM If you agree I can commit the necessary change(s). I also found a few indenting issues which I would also commit (probable in a separate commit).

**Comment by Albrecht-S** [comment]
> As for the other question, I noticed that the difference is only whether it is printed or not, and it will cause pointer errors in the next loop. The same goes for testing.

@xuyun018 Please see my comment above. I don't think that there is a "next loop" for printing. Do you agree with my statements?

**Comment by xuyun018** [comment]
![1](https://github.com/user-attachments/assets/f6c49961-257a-4feb-b2c7-9eeb7c6b0c04)
I used this image for testing, and its height is 695 pixels. Because the static variable inside is limited to 256K, the variable "blocking" is now 130 pixels. When printing, the buffer will be NULL in the second loop. I tested it on my end, but the program crashed.

**Comment by Albrecht-S** [comment]
@xuyun018 Thank you very much for posting the image. Which program did you use for printing? One of the FLTK test programs or one of your own? Can you post instructions or a short demo so we can use the same code for testing? TIA

**Comment by xuyun018** [comment]
class ImageWindow : public Fl_Window {
public:
    Fl_JPEG_Image* image;
    ImageWindow(int W, int H, const char* title = 0)
        : Fl_Window(W, H, title), image(nullptr) {
        // 尝试加载JPEG图像  
        image = new Fl_JPEG_Image("1.jpg");

        if (!image) {  
            end();
        }

        new Fl_Box(0, 0, w(), h(), "JPEG Image will be here");
    }

    ~ImageWindow() {
        delete image;
    }

    void draw() FL_OVERRIDE;
};

void ImageWindow::draw()
{
    Fl_RGB_Image* pimage;

    image->draw(0, 0);
}

I don't have a printer, so I moved the code below and the program crashed, so I learned that if there was a printer, it would also crash.
      delete[] buffer;
      buffer = NULL;
      buffer_size = 0;

**Comment by xuyun018** [comment]
Because I need to create animations on Widgets, I have high performance requirements. Therefore, I have added several interfaces to FLTK's fl_draw unit for direct operation, reducing data conversion and memory usage. I have already written it for Windows, and now I am writing the Linux version

**Comment by Albrecht-S** [comment]
@xuyun018 You wrote:
> I don't have a printer, so I moved the code below and the program crashed, so I learned that if there was a printer, it would also crash.

Sorry, I don't understand this statement. If you "moved the code below" - I take it this means **_you patched the FLTK code_** - then this is very likely the cause of the crash. I analyzed the original code and came to the conclusion that deleting the buffer inside the block started [here](https://github.com/fltk/fltk/blob/68f437b63e9e54769b1854097f2c95dd1cce12b7/src/drivers/GDI/Fl_GDI_Graphics_Driver_image.cxx#L255) is correct, but if you moved it **outside this block** it can cause the **crash for large images**.

See code at https://github.com/fltk/fltk/blob/68f437b63e9e54769b1854097f2c95dd1cce12b7/src/drivers/GDI/Fl_GDI_Graphics_Driver_image.cxx#L174 and https://github.com/fltk/fltk/blob/68f437b63e9e54769b1854097f2c95dd1cce12b7/src/drivers/GDI/Fl_GDI_Graphics_Driver_image.cxx#L177 where "blocking" is disabled for "printing", in other words: blocking is only allowed if NOT printing...

**Comment by Albrecht-S** [comment]
FWIW, here is a full demo program that does NOT show a problem. Note that this program does NOT use printing but it may be a start for a **_real_** test program.

[issue_1033.cxx.txt](https://github.com/user-attachments/files/16481050/issue_1033.cxx.txt)

**Comment by xuyun018** [comment]
I didn't fix the original code, what I meant was adding code, another project. 

Sorry, I didn't see the logic above,

**Comment by xuyun018** [comment]
I moved the code to verify, it was my mistake, I didn't see the logic of the printer's judgment above.

**Comment by Albrecht-S** [comment]
@xuyun018 OK, thanks for confirmation.

@MatthiasWM I assigned myself and I'll remove the obsolete statement and fix indenting etc..

FWIW: following is a full demo program that prints the large image posted above. The code for printing was taken from the docs and I tested it on `Linux` using `Wine`. I verified that the buffer was indeed deleted w/o issues.

[issue_1033.cxx.txt](https://github.com/user-attachments/files/16481148/issue_1033.cxx.txt)
